File: sesa-backend/src/main/java/com/sesa/salud/service/EvolucionService.java
/**
 * Autor: Ing. J Sebastian Vargas S
 */

package com.sesa.salud.service;

import com.sesa.salud.dto.EvolucionDto;
import com.sesa.salud.dto.EvolucionRequestDto;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface EvolucionService {

    List<EvolucionDto> findByAtencionId(Long atencionId, Pageable pageable);

    /** Todas las evoluciones del paciente (desde atenciones de su historia clínica, ej. urgencias). */
    List<EvolucionDto> findByPacienteId(Long pacienteId, Pageable pageable);

    EvolucionDto findById(Long id);

    EvolucionDto create(EvolucionRequestDto dto);

    EvolucionDto update(Long id, EvolucionRequestDto dto);
}

File: sesa-salud/src/app/core/services/evolucion.service.ts
/**
 * Servicio de Evoluciones médicas — registro de evolución clínica en urgencias/consultas.
 * Autor: Ing. J Sebastian Vargas S
 */

import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface EvolucionDto {
  id: number;
  atencionId: number;
  notaEvolucion: string;
  fecha: string;
  profesionalId?: number;
  profesionalNombre?: string;
  createdAt?: string;
}

export interface EvolucionRequestDto {
  atencionId: number;
  notaEvolucion: string;
  fecha?: string;
  profesionalId?: number;
}

@Injectable({ providedIn: 'root' })
export class EvolucionService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiUrl}/evoluciones`;

  listarPorAtencion(atencionId: number): Observable<EvolucionDto[]> {
    const params = new HttpParams()
      .set('atencionId', atencionId.toString())
      .set('size', '100');
    return this.http.get<EvolucionDto[]>(this.apiUrl, { params });
  }

  /** Evoluciones del paciente (urgencias y otras atenciones) para timeline en Historia Clínica. */
  listarPorPaciente(pacienteId: number): Observable<EvolucionDto[]> {
    return this.http.get<EvolucionDto[]>(`${this.apiUrl}/paciente/${pacienteId}`, {
      params: new HttpParams().set('size', '100')
    });
  }

  crear(dto: EvolucionRequestDto): Observable<EvolucionDto> {
    return this.http.post<EvolucionDto>(this.apiUrl, dto);
  }

  actualizar(id: number, dto: Partial<EvolucionRequestDto>): Observable<EvolucionDto> {
    return this.http.put<EvolucionDto>(`${this.apiUrl}/${id}`, dto);
  }
}
